misk-proto/src/main/proto/status.proto

```diff
@@ -0,0 +1,92 @@
+
+package google.rpc;
+
+import "google/protobuf/any.proto";
+
+option go_package = "google.golang.org/genproto/googleapis/rpc/status;status";
+option java_multiple_files = true;
+option java_outer_classname = "StatusProto";
+option java_package = "com.google.rpc";
+option objc_class_prefix = "RPC";
+
+
+// The `Status` type defines a logical error model that is suitable for different
+// programming environments, including REST APIs and RPC APIs. It is used by
+// [gRPC](https://github.com/grpc). The error model is designed to be:
+//
+// - Simple to use and understand for most users
+// - Flexible enough to meet unexpected needs
+//
+// # Overview
+//
+// The `Status` message contains three pieces of data: error code, error message,
+// and error details. The error code should be an enum value of
+// [google.rpc.Code][google.rpc.Code], but it may accept additional error codes if needed.  The
+// error message should be a developer-facing English message that helps
+// developers *understand* and *resolve* the error. If a localized user-facing
+// error message is needed, put the localized message in the error details or
+// localize it in the client. The optional error details may contain arbitrary
+// information about the error. There is a predefined set of error detail types
+// in the package `google.rpc` which can be used for common error conditions.
+//
+// # Language mapping
+//
+// The `Status` message is the logical representation of the error model, but it
+// is not necessarily the actual wire format. When the `Status` message is
+// exposed in different client libraries and different wire protocols, it can be
+// mapped differently. For example, it will likely be mapped to some exceptions
+// in Java, but more likely mapped to some error codes in C.
+//
+// # Other uses
+//
+// The error model and the `Status` message can be used in a variety of
+// environments, either with or without APIs, to provide a
+// consistent developer experience across different environments.
+//
+// Example uses of this error model include:
+//
+// - Partial errors. If a service needs to return partial errors to the client,
+//     it may embed the `Status` in the normal response to indicate the partial
+//     errors.
+//
+// - Workflow errors. A typical workflow has multiple steps. Each step may
+//     have a `Status` message for error reporting purpose.
+//
+// - Batch operations. If a client uses batch request and batch response, the
+//     `Status` message should be used directly inside batch response, one for
+//     each error sub-response.
+//
+// - Asynchronous operations. If an API call embeds asynchronous operation
+//     results in its response, the status of those operations should be
+//     represented directly using the `Status` message.
+//
+// - Logging. If some API errors are stored in logs, the message `Status` could
+//     be used directly after any stripping needed for security/privacy reasons.
+message Status {
```

Comment:
Do you have more context around why we would ship a copy of `Status`?

Reply:
Sure! Including a full un-padded-base64-proto-encoded copy of Status as a `grpc-status-details-bin` trailer is the common way to communicate rich errors to clients in gRPC. Unfortunately, this [detail](https://github.com/grpc/grpc/issues/24007) didn't make it in the spec (yet).

It's widely supported in practice. [grpc-go](https://github.com/grpc/grpc-go/blob/23a83dd097ec07fc7ddfb4a30c675763e4972ba4/internal/transport/handler_server.go#L212) has it, [grpc-java](https://github.com/grpc/grpc-java/blob/d4fa0ecc07495097453b0a2848765f076b9e714c/protobuf/src/main/java/io/grpc/protobuf/StatusProto.java#L37), as well as [Python, Ruby and C++](https://github.com/grpc/grpc/search?q=grpc-status-details-bin) and virtually every compatible implementation.

It lets servers return errors using the standard `Status` proto which allows embedding custom error details (for example [error_details.proto](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto)). In Misk we do so by throwing a `WebApiException`. To keep the API compatible with Wire-generated messages, `WebApiException` accepts a list of `AnyMessage`—a Wire [runtime](https://square.github.io/wire/3.x/wire-runtime/com.squareup.wire/-any-message/). This makes it easy to use but it means we need to compile our own copy of `status.proto`.